test-gui/src/package/view-decoded-linear-position-plot/DecodedLinearPositionDrawing.ts

```diff
@@ -0,0 +1,212 @@
+    canvasTargetHeight: number
+    painter: OffscreenPainter
+    linesRepresentation: TimeColumn[]
+    scale: number
+    downsampledRangeStart: number
+    downsampledRangeEnd: number
+}
+// Given a Canvas with stuff drawn on it and a data set, and a desired range of data from the data set,
+// return the x-coordinate pixel range in the offscreen canvas that contains the requested data range.
+export const useOffscreenCanvasRange = (props: OffscreenRenderProps): [number, number] => {
+    const contentsStart = useRef<number>(0)
+    const contentsEnd = useRef<number>(0)
+    const currentScale = useRef<number>(0)
+    const { canvas, canvasTargetWidth, canvasTargetHeight, painter, scale, linesRepresentation, downsampledRangeStart, downsampledRangeEnd } = props
+
+    if (canvas === undefined) return [0, 0]
+
+    // Checking this on a useEffect hook was not responsive enough--it wasn't updating until the *second* rerender.
+    if (scale !== currentScale.current) {
+        currentScale.current = scale
+        contentsStart.current = 0
+        contentsEnd.current = 0
+    }
+
+    // NOTE: React wants to resize this canvas on every soft reload. I don't expect it to be a problem generally, but
+    // when it does happen we need to invalidate the cache since the canvas contents get cleared by the resize.
+    if ((canvas.width !== canvasTargetWidth || canvas.height !== canvasTargetHeight)) {
+        canvas.width = canvasTargetWidth
+        canvas.height = canvasTargetHeight
+        contentsStart.current = 0
+        contentsEnd.current = 0
+    }
+
+    console.assert(contentsStart.current <= contentsEnd.current)
+    if ((downsampledRangeEnd - downsampledRangeStart) > (canvas.width)) {
+        throw Error(`Impossible situation: requested window ${downsampledRangeStart}-${downsampledRangeEnd} does not fit in canvas width ${canvas.width} as allowed by current scale factor ${scale}`)
+    }
+
+    // Request can be served from cache--do so
+    if (contentsStart.current <= downsampledRangeStart && downsampledRangeEnd <= contentsEnd.current) {
+        const pixelStart = (downsampledRangeStart - contentsStart.current)
+        const pixelEnd = (downsampledRangeEnd - contentsStart.current)
+        return [pixelStart, pixelEnd]
+    }
+
+    // Request cannot be served from cache.
+    const visibleRangeMidpoint = Math.floor((downsampledRangeEnd - downsampledRangeStart)/2) + downsampledRangeStart
+    const {targetStart, targetEnd} = getRenderTargetRange(visibleRangeMidpoint, linesRepresentation.length - 1, canvas.width)
+
+    // TODO: Can this be passed off to a separate worker thread?
+    // TODO: Further refactoring could separate this entirely, since the rest of this function does not care about a)
+    // whether the canvas actually exists and b) the painter, canvas, or lines-representation. Mostly.
+    updateCachedImage(targetStart, targetEnd, contentsStart.current, contentsEnd.current, painter, linesRepresentation, canvas)
```

Comment:
This is one of the more expensive operations, though honestly not nearly as bad (at this point) as downsampling or re-representing the data. I'm used to these values not exceeding maybe 300 ms at the outside.